content/post/review-building-static-analysis-tools.md

```diff
@@ -0,0 +1,357 @@
+  This by necessity constraints the scope of a possible analysis and a tooling required
+
+* It also means that repairing programs is important. For the discussion on tooling available for code
+  transformations see **Technical Details** section below.
+
+Learning such modifications from examples, instead of manual coding by engineers is also a bleeding edge
+research topic [https://github.com/KTH/learning4repair](https://github.com/KTH/learning4repair).
+
+### 2. Best way to **present the results of analysis**
+> Developers were more likely to fix bugs if presented with analysis results early and as part of their normal workflow
+
+This immediately implies that **reporting issues sooner is better.**
+
+It also leads to the conclusion that the best bet is to integrate checks either ***directly into compilers*** -
+familiar tools on whose feedback as errors and warnings developers are already relaying day to day. Or, if
+that is not possible, _**code review** is a good time_ for new changes — before they are committed to the
+version control system.
+
+Criteria that must hold for ***compile time*** checks:
+
+* 0 false positives
+
+* easy to understand, actionable and easy to fix
+
+* report issues affecting only correctness, rather than style or best practices
+
+Criteria that must hold for ***code review time*** checks:
+
+* <10% false positives
+
+* be understandable, actionable and easy to fix
+
+* have the potential for significant impact on code quality
+
+Last but not least, enabling any new *compile time* check can suddenly brake a developer working on existing
+code, which is clearly unacceptable. In order to avoid that, first **a tooling for large-scale code modifications**
+must be run over the whole codebase, before moving those checks to compile time for the rest of the company.
+
+And those are ClangMR and JavacFlume — projects that are only briefly mentioned in this insightful paper.
+
+*That is it, thank you for reading. We will post more on papers in this field soon.*
+
+{{% center %}} … {{% /center %}}
+
+Now I will take a liberty and cover a few technical details that were not in the scope of the original paper
+but are very related, in order to help us see a bigger picture. Those were described in other places by other
+employees of the same company.
+
+## Technical details
+
+Based on the internal success-story of C++ with [ClangMR tool](https://research.google.com/pubs/pub41342.html)
+for matching/traversing/transforming Abstract Syntax Tree (AST) at scale, a similar tooling was built for Java.
+
+{{% youtube ZpvvmvITOrk %}}
+
+Project [Error-Prone](https://github.com/google/error-prone) is a compiler extension that is able to perform
+arbitrary analysis on the *fully typed AST*. One thing to notice is that one can not get such input by using
+only a parser even as advanced as [https://doc.bblf.sh](https://doc.bblf.sh/). Running a full build would be
```

Comment:
use the name of the project, not the link itself `[babelfish](https://doc.bblf.sh)`

Same for all other links